bundles/org.openhab.binding.upnpcontrol/src/main/java/org/openhab/binding/upnpcontrol/internal/audiosink/UpnpAudioSink.java

```diff
@@ -65,27 +63,32 @@ public String getId() {
     }
 
     @Override
-    public void process(@Nullable AudioStream audioStream)
+    protected void processAsynchronously(@Nullable AudioStream audioStream)
             throws UnsupportedAudioFormatException, UnsupportedAudioStreamException {
         if (audioStream == null) {
             stopMedia();
             return;
         }
 
-        String url = null;
-        if (audioStream instanceof URLAudioStream) {
-            URLAudioStream urlAudioStream = (URLAudioStream) audioStream;
-            url = urlAudioStream.getURL();
+        if (audioStream instanceof URLAudioStream urlAudioStream) {
+            playMedia(urlAudioStream.getURL());
         } else if (!callbackUrl.isEmpty()) {
-            String relativeUrl = audioStream instanceof FixedLengthAudioStream
-                    ? audioHTTPServer.serve((FixedLengthAudioStream) audioStream, 20)
-                    : audioHTTPServer.serve(audioStream);
-            url = String.valueOf(this.callbackUrl) + relativeUrl;
+            StreamServed streamServed;
+            try {
+                streamServed = audioHTTPServer.serve(audioStream, 20, true);
+            } catch (IOException e) {
+                throw new UnsupportedAudioStreamException(handler.getUDN() + " can only handle clonable audio streams.",
+                        audioStream.getClass(), e);
+            }
+            streamServed.playEnd().thenRun(() -> this.playbackFinished(audioStream));
+            playMedia(callbackUrl + streamServed.url());
         } else {
             logger.warn("We do not have any callback url, so {} cannot play the audio stream!", handler.getUDN());
-            return;
         }
-        playMedia(url);
+        try {
+            audioStream.close();
```

Comment:
@dalgwen : please confirm that it is ok to close the audio stream at this place (async handling).

Reply:
You can, but I think it's better not to :

- You can do this, because your AudioSink asks the servlet to make it a multi time stream. 
And thus, all call to the servlet will clone the stream, and the original stream won't be needed. So it's OK if you close it.

- You don't have to, because the AudioServlet will close it (when removeTimedOutStreams or createClonableInputStream is called). And since the rewrite, removeTimedOutStreams is always called (before, it was not and it was a potential resource leak)

- Please also note that for Audio Sink that do not require multi time stream, this is a faulty behaviour. A one time stream is directly used by the servlet to give audio data to the requesting device, and thus should not be closed.

So I suggest not closing it, to have a consistent code.

Reply:
Ok, fine.

I will keep this only for the case URLAudioStream.